src/os/bluestore/BlueStore.h

```diff
@@ -1131,208 +1235,137 @@ class BlueStore : public ObjectStore,
+    void _add(Buffer *b, int level, Buffer *near) override {
       if (near) {
-	auto q = buffer_lru.iterator_to(*near);
-	buffer_lru.insert(q, *b);
+        auto q = lru.iterator_to(*near);
+        lru.insert(q, *b);
       } else if (level > 0) {
-	buffer_lru.push_front(*b);
+        lru.push_front(*b);
       } else {
-	buffer_lru.push_back(*b);
-      }
-      buffer_size += b->length;
-    }
-    void _rm_buffer(Buffer *b) override {
-      ceph_assert(buffer_size >= b->length);
-      buffer_size -= b->length;
-      auto q = buffer_lru.iterator_to(*b);
-      buffer_lru.erase(q);
-    }
-    void _move_buffer(Cache *src, Buffer *b) override {
-      src->_rm_buffer(b);
-      _add_buffer(b, 0, nullptr);
-    }
-    void _adjust_buffer_size(Buffer *b, int64_t delta) override {
-      ceph_assert((int64_t)buffer_size + delta >= 0);
-      buffer_size += delta;
-    }
-    void _touch_buffer(Buffer *b) override {
-      auto p = buffer_lru.iterator_to(*b);
-      buffer_lru.erase(p);
-      buffer_lru.push_front(*b);
+        lru.push_back(*b);
+      }
+      buffer_bytes += b->length;
+      num = lru.size();
+    }
+    void _rm(Buffer *b) override {
+      ceph_assert(buffer_bytes >= b->length);
+      buffer_bytes -= b->length;
+      auto q = lru.iterator_to(*b);
+      lru.erase(q);
+      num = lru.size();
+    }
+    void _move(BufferCacheShard *src, Buffer *b) override {
+      src->_rm(b);
+      _add(b, 0, nullptr);
+    }
+    void _adjust_size(Buffer *b, int64_t delta) override {
+      ceph_assert((int64_t)buffer_bytes + delta >= 0);
+      buffer_bytes += delta;
+    }
+    void _touch(Buffer *b) override {
+      auto p = lru.iterator_to(*b);
+      lru.erase(p);
+      lru.push_front(*b);
+      num = lru.size();
       _audit("_touch_buffer end");
     }
 
-    void _trim(uint64_t onode_max, uint64_t buffer_max) override;
+    void _trim_to(uint64_t max) override;
 
-    void add_stats(uint64_t *onodes, uint64_t *extents,
-		   uint64_t *blobs,
-		   uint64_t *buffers,
-		   uint64_t *bytes) override {
-      std::lock_guard l(lock);
-      *onodes += onode_lru.size();
+    void add_stats(uint64_t *extents,
+                   uint64_t *blobs,
+                   uint64_t *buffers,
+                   uint64_t *bytes) override {
       *extents += num_extents;
       *blobs += num_blobs;
-      *buffers += buffer_lru.size();
-      *bytes += buffer_size;
+      *buffers += num;
+      *bytes += buffer_bytes;
     }
-
 #ifdef DEBUG_CACHE
     void _audit(const char *s) override;
 #endif
   };
 
-  // 2Q cache for buffers, LRU for onodes
-  struct TwoQCache : public Cache {
-  private:
-    // stick with LRU for onodes for now (fixme?)
-    typedef boost::intrusive::list<
-      Onode,
-      boost::intrusive::member_hook<
-        Onode,
-	boost::intrusive::list_member_hook<>,
-	&Onode::lru_item> > onode_lru_list_t;
+  /// A TwoQ buffer Cache Shard
+  struct TwoQBufferCacheShard : public BufferCacheShard {
```

Comment:
can we move the different implementations of `BufferCacheShard` into .cc ? as they are not part of the interface of BlueStore.